ghcide/src/Development/IDE/Types/Logger.hs

```diff
@@ -57,3 +91,180 @@ logTelemetry x = logPriority x Telemetry
 
 noLogging :: Logger
 noLogging = Logger $ \_ _ -> return ()
+
+data WithPriority a = WithPriority { priority :: Priority, payload :: a } deriving Functor
+
+-- | Note that this is logging actions _of the program_, not of the user.
+--   You shouldn't call warning/error if the user has caused an error, only
+--   if our code has gone wrong and is itself erroneous (e.g. we threw an exception).
+data Recorder msg = Recorder
+  { logger_ :: forall m. (HasCallStack, MonadIO m) => msg -> m () }
+
+logWith :: (HasCallStack, MonadIO m) => Recorder (WithPriority msg) -> Priority -> msg -> m ()
+logWith recorder priority msg = withFrozenCallStack $ logger_ recorder (WithPriority priority msg)
```

Comment:
I feel like I'm missing something: where does the call stack get used?

Reply:
It never gets used, but it was in the original Pepe example, and at one point I used it to print the source location of the log call in the logs, but since things are different now not sure if that's still possible. I'll remove it.

Reply:
Oh no, I find it really useful to print call sites in log lines. What makes it not possible anymore?

Reply:
OK back in it goes. I added a `callStack_ :: CallStack` field to `WithPriority` to do it easily.

I think I didn't remember correctly. Wasn't using `WithPriority` in the past, so was trying to create a chain of `HasCallStacks` from call site to output without storing a `CallStack`, but there was missing link somewhere, and I didn't spend any time to figure it out. So after `WithPriority` was added I got confused about why I removed it.